base/compiler/abstractinterpretation.jl

```diff
@@ -860,17 +860,15 @@ function abstract_iteration(interp::AbstractInterpreter, @nospecialize(itft), @n
     end
     # From here on, we start asking for results on the widened types, rather than
     # the precise (potentially const) state type
-    statetype = widenconst(statetype)
     valtype = widenconst(valtype)
     while valtype !== Any
-        stateordonet = abstract_call_known(interp, iteratef, nothing, Any[Const(iteratef), itertype, statetype], sv).rt
```

Comment:
This is the only line that should be moved, though I think we can rearrange some others a bit to sometimes skip work that is unnecessary.
```diff
diff --git a/base/compiler/abstractinterpretation.jl b/base/compiler/abstractinterpretation.jl
index 296a30cdb2..0da2ce68b3 100644
--- a/base/compiler/abstractinterpretation.jl
+++ b/base/compiler/abstractinterpretation.jl
@@ -860,25 +860,28 @@ function abstract_iteration(interp::AbstractInterpreter, @nospecialize(itft), @n
     end
     # From here on, we start asking for results on the widened types, rather than
     # the precise (potentially const) state type
-    statetype = widenconst(statetype)
     valtype = widenconst(valtype)
-    while valtype !== Any
-        stateordonet = abstract_call_known(interp, iteratef, nothing, Any[Const(iteratef), itertype, statetype], sv).rt
-        stateordonet = widenconst(stateordonet)
-        nounion = typesubtract(stateordonet, Nothing, 0)
-        if !isa(nounion, DataType) || !(nounion <: Tuple) || isvatuple(nounion) || length(nounion.parameters) != 2
-            valtype = Any
-            break
-        end
-        if nounion.parameters[1] <: valtype && nounion.parameters[2] <: statetype
-            if typeintersect(stateordonet, Nothing) === Union{}
-                # Reached a fixpoint, but Nothing is not possible => iterator is infinite or failing
-                return Any[Bottom], nothing
+    statetype = widenconst(statetype)
+    if valtype !== Any
+        while true
+            stateordonet = widenconst(stateordonet)
+            nounion = typesubtract(stateordonet, Nothing, 0)
+            if !isa(nounion, DataType) || !(nounion <: Tuple) || isvatuple(nounion) || length(nounion.parameters) != 2
+                valtype = Any
+                break
             end
-            break
+            if nounion.parameters[1] <: valtype && nounion.parameters[2] <: statetype
+                if typeintersect(stateordonet, Nothing) === Union{}
+                    # Reached a fixpoint, but Nothing is not possible => iterator is infinite or failing
+                    return Any[Bottom], nothing
+                end
+                break
+            end
+            valtype = tmerge(valtype, nounion.parameters[1])
+            valtype === Any && break
+            statetype = tmerge(statetype, nounion.parameters[2])
+            stateordonet = abstract_call_known(interp, iteratef, nothing, Any[Const(iteratef), itertype, statetype], sv).rt
         end
-        valtype = tmerge(valtype, nounion.parameters[1])
-        statetype = tmerge(statetype, nounion.parameters[2])
     end
     push!(ret, Vararg{valtype})
     return ret, nothing
```

Reply:
No, that leads to the following unsoundness:
```julia
julia> f() = (1:40...,)
f (generic function with 1 method)

julia> @code_typed f()
CodeInfo(
1 ─     Core._apply_iterate(Base.iterate, Core.tuple, $(QuoteNode(1:40)))::Union{}
└──     unreachable
) => Union{}
```
In this case, the first loop is terminates because `MAX_TUPLE_SPLAT` is exceeded. Then `statetype`, `valtype` and `stateordonet` all are `Const`s. Next we widen them to `Int`, `Int`, and `Tuple{Int,Int}`, respectively. So `nounion == Tuple{Int,Int}` also thus `nounion.parameters[1] <: valtype && nounion.parameters[2] <: statetype`. But at the same time, `typeintersect(stateordonet, Nothing) === Union{}`, so a `nothing`-free fixpoint is declared erroneously. That's why in my version, I don't `widenconst` the `statetype` and use `⊑`, so that in the first iteration `nounion.parameters[2] ⊑ statetype == Int ⊑ Const(n) == false`. In successive iterations, `statetype` will always be a proper type, so we could use `<:` then, but the overhead of `⊑` in the both-arguments-are-types case seems negligible here so I chose not to worry about it.

I think somewhere in the `abstractarray` tests we do a splat of a looong range which would have caught this, but I'll add a more direct test while at it.

Reply:
Ah, that is true. I marked it as a bugfix, since it appeared to me that we might end up skipping including the first `valtype` (from `iterate(foo)`) into the final result, as we begin this loop with discarding that result and computing `iterate(foo, state)`

Reply:
> I marked it as a bugfix, since it appeared to me that we might end up skipping including the first valtype (from iterate(foo)) into the final result

...but we basically repeated that `iterate` (with widened, but not updated state), so this was indeed not a correctness issue ... right?

Reply:
yeah, mostly we do repeat the same, but it appears to get specifically confused by that particular case, but perhaps it doesn't get this wrong. Since we start with statetype===Bottom, it will conclude that the result is Bottom, and then luckily we will replace that with Any, but it is not great.

```
iterate(::Foo) = rand(Bool) ? nothing : (nothing, 1)
iterate(::Foo, i) = i < 16 ? (i, i + 1) : nothing
```